### PR TITLE
fix(reflection): MetaData.reflect() raises exception with sqlalchemy v2.0.36

### LINKED ISSUE #1128: [Bug]: MetaData.reflect() raises exception with sqlalchemy v2.0.36
### What happened?

This [commit](https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5519/3/lib/sqlalchemy/dialects/postgresql/base.py) in sqlalchemy v2.0.36 causes a regression when creating a table in DuckDB. A SQLAlchemy developer [says](https://github.com/sqlalchemy/sqlalchemy/discussions/12011#discussioncomment-10992396) that this package will need to override the new statement.

```python
from sqlalchemy import MetaData, create_engine, text

engine = create_engine("duckdb:///:memory:")
metadata = MetaData()
with engine.connect() as conn:
    conn.execute(text("CREATE TABLE tbl(col1 INTEGER)"))
    conn.commit()
metadata.reflect(engine)
```

```
/private/tmp/env/lib/python3.11/site-packages/duckdb_engine/__init__.py:174: DuckDBEngineWarning: duckdb-engine doesn't yet support reflection on indices
  warnings.warn(
Traceback (most recent call last):
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
    cursor.execute(statement, parameters)
  File "/private/tmp/env/lib/python3.11/site-packages/duckdb_engine/__init__.py", line 140, in execute
    self.__c.execute(statement, parameters)
duckdb.duckdb.CatalogException: Catalog Error: Type with name REGCLASS does not exist!

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/dthom/repos/chronify/scripts/repro_alchemy.py", line 8, in <module>
    metadata.reflect(engine)
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/sql/schema.py", line 5828, in reflect
    _reflect_info = insp._get_reflection_info(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/reflection.py", line 2024, in _get_reflection_info
    table_comment=run(self.get_multi_table_comment, optional=True),
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/reflection.py", line 2001, in run
    res = meth(filter_names=_fn, **kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/reflection.py", line 1377, in get_multi_table_comment
    self.dialect.get_multi_table_comment(
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/base.py", line 4709, in get_multi_table_comment
    result = connection.execute(query, params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2355, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/private/tmp/env/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 941, in do_execute
    cursor.execute(statement, parameters)
  File "/private/tmp/env/lib/python3.11/site-packages/duckdb_engine/__init__.py", line 140, in execute
    self.__c.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (duckdb.duckdb.CatalogException) Catalog Error: Type with name REGCLASS does not exist!
[SQL: SELECT pg_catalog.pg_class.relname, pg_catalog.pg_description.description 
FROM pg_catalog.pg_class LEFT OUTER JOIN pg_catalog.pg_description ON pg_catalog.pg_class.oid = pg_catalog.pg_description.objoid AND pg_catalog.pg_description.objsubid = $1 AND pg_catalog.pg_description.classoid = CAST($2 AS REGCLASS) JOIN pg_catalog.pg_namespace ON pg_catalog.pg_namespace.oid = pg_catalog.pg_class.relnamespace 
WHERE pg_catalog.pg_class.relkind = ANY (ARRAY[$3, $4, $5]) AND pg_catalog.pg_table_is_visible(pg_catalog.pg_class.oid) AND pg_catalog.pg_namespace.nspname != $6 AND pg_catalog.pg_class.relname IN ($7)]
[parameters: (0, 'pg_catalog.pg_class', 'r', 'p', 'f', 'pg_catalog', 'tbl')]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

### DuckDB Engine Version

0.13.2

### DuckDB Version

1.1.2

### SQLAlchemy Version

2.0.36

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Aarya2004**
Fixes #1128 using the suggestion from [this discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/12011) to override this [commit](https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5519/3/lib/sqlalchemy/dialects/postgresql/base.py). Function was just overridden and a small test case created to test this case.

### CODE REVIEW

File: duckdb_engine/tests/test_reflection.py
Comment (NickCrews) [inline duckdb_engine/tests/test_reflection.py:3]: 1. you need to put this inside a test function for pytest to pick it up.
2. I would probably put this inside test_basic.py instead of making a whole new file.

so, would look something like

```
# test_basic.py
# other tests up here...

def test_reflection(engine: Engine) -> None:
    from sqlalchemy import MetaData

    metadata = MetaData()

    with engine.connect() as conn:
        conn.execute(text("CREATE TABLE tbl(col1 INTEGER)"))
        conn.commit()
    metadata.reflect(engine)
```

and then you would run this with something like `poetry run pytest -k reflection duckdb_engine/tests/test_basic.py`

File: duckdb_engine/__init__.py
Comment (NickCrews) [inline duckdb_engine/__init__.py:28]: tests make it look like you are importing from the wrong place? perhaps it should be

`from sqlalchemy.dialects.postgresql import pg_catalog`?

See https://github.com/sqlalchemy/sqlalchemy/blob/7b12826a02fa1f83acd64bd531c9e47832d22cb0/lib/sqlalchemy/dialects/postgresql/pg_catalog.py

File: duckdb_engine/__init__.py
Comment (NickCrews) [inline duckdb_engine/__init__.py:607]: I would add a comment here saying why we need this, eg
`# fix for https://github.com/Mause/duckdb_engine/issues/1128`

File: duckdb_engine/tests/test_reflection.py
Comment (Aarya2004) [reply duckdb_engine/tests/test_reflection.py:3]: Sounds good, I'll make the update. Thanks for the tip!

File: duckdb_engine/__init__.py
Comment (Aarya2004) [reply duckdb_engine/__init__.py:608]: Done in the latest commit!

File: duckdb_engine/__init__.py
Comment (Aarya2004) [reply duckdb_engine/__init__.py]: Done in the latest commit!

File: duckdb_engine/tests/test_reflection.py
Comment (Aarya2004) [reply duckdb_engine/tests/test_reflection.py]: Done in the latest commit!

**Comment by flaviomartins** [comment]
@NickCrews is this PR ready for merging?

File: duckdb_engine/__init__.py
Comment (NickCrews) [inline duckdb_engine/__init__.py:610]: in [the upstream code](https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5519/3/lib/sqlalchemy/dialects/postgresql/base.py#4677) they have a cache on this call. Can you add that here too? IDK if it's really needed, but it sure is easy:

```
import functools

...

@functools.cache()
def _comment_query...
```

**Review by NickCrews — CHANGES_REQUESTED** [review]
I'd love two adjustments, and then if CI passes then this looks great, thanks for your work on this

**Comment by NickCrews** [comment]
Once I'm happy this also will need approval from @Mause, I'm just a rando, she is the maintainer.

File: duckdb_engine/tests/test_basic.py
Comment (Aarya2004) [reply duckdb_engine/tests/test_basic.py]: Fixed in the latest commit.

File: duckdb_engine/__init__.py
Comment (Aarya2004) [reply duckdb_engine/__init__.py:612]: Fixed in the latest commit. I added lru_cache since that's what is used in the upstream code and cache was giving me errors but I did read online that they're similar functions. Is that alright?

**Comment by Aarya2004** [comment]
Sounds good, thanks so much for reviewing our code @NickCrews! We've just made the fixes that you pointed out. Will ask Mause for a review and approval once you've confirmed the code looks good.

**Comment by NickCrews** [comment]
Why did you make the change to getattr()?

Besides that LGTM!

**Comment by Aarya2004** [comment]
> Why did you make the change to getattr()?

Just wanted to be consistent with the way we were calling the upstream functions. I realized that functions that weren't overridden, we call them using that method (as I did with the _kinds_to_relkinds func).

TLDR - just to be more consistent with how we called functions that aren't overridden.

File: duckdb_engine/__init__.py
Comment (NickCrews) [reply duckdb_engine/__init__.py:612]: Yes, without a cache size I think they are the same

**Comment by NickCrews** [comment]
Ok sounds good, @Mause I'm happy with this whenever you get a chance to review, thanks!

**Comment by Aarya2004** [comment]
Thanks for all the help @NickCrews !

**Comment by Aarya2004** [comment]
Hello @Mause, just wanted to follow up on this!

**Comment by Mause** [comment]
Can you look into the failing tests please?

**Comment by Aarya2004** [comment]
Looks like the runs failed due to: ImportError: cannot import name 'pg_catalog' from 'sqlalchemy.dialects.postgresql'
I've added a line to make sure the changes only apply to sqlalchemy versions >= 2.0.36, so the runs should be good now!

**Comment by Mause** [comment]
Thanks!!

**Pull request merged**
